File: macos/Sources/VotProto.swift
import Foundation

enum VotProto {
    struct TranslationResponse {
        var url = ""
        var status = 0
        var remainingTime = 0
        var translationId = ""
        var language = ""
        var message = ""
        var isLivelyVoice = false
    }

    struct SessionResponse {
        let secretKey: String
        let expires: Int
    }

    enum ProtoError: LocalizedError {
        case malformed(String)

        var errorDescription: String? {
            switch self {
            case .malformed(let message): return message
            }
        }
    }

    static func encodeSessionRequest(uuid: String, module: String) -> Data {
        var writer = Writer()
        writer.string(1, uuid)
        writer.string(2, module)
        return writer.data
    }

    static func decodeSessionResponse(_ data: Data) throws -> SessionResponse {
        var reader = Reader(data)
        var secretKey = ""
        var expires = 0

        while reader.hasRemaining {
            guard let tag = try reader.readTag() else { break }
            switch tag.fieldNumber {
            case 1: secretKey = try reader.readString(wireType: tag.wireType)
            case 2: expires = try reader.readInt32(wireType: tag.wireType)
            default: try reader.skip(wireType: tag.wireType)
            }
        }

        guard !secretKey.isEmpty else {
            throw ProtoError.malformed("VOT session response did not include a secret key")
        }
        return SessionResponse(secretKey: secretKey, expires: expires)
    }

    static func encodeTranslationRequest(
        url: String,
        firstRequest: Bool,
        duration: Double,
        language: String,
        responseLanguage: String,
        videoTitle: String,
        useLivelyVoice: Bool
    ) -> Data {
        var writer = Writer()
        writer.string(3, url)
        writer.bool(5, firstRequest)
        writer.double(6, duration)
        writer.int32(7, 1)
        writer.string(8, language)
        writer.string(14, responseLanguage)
        writer.int32(15, 1)
        writer.int32(16, 2)
        writer.bool(18, useLivelyVoice)
        if !videoTitle.isEmpty { writer.string(19, videoTitle) }
        return writer.data
    }

    static func encodeAudioRequest(translationId: String, url: String, fileId: String) -> Data {
        var audioInfo = Writer()
        audioInfo.string(1, fileId)

        var writer = Writer()
        writer.string(1, translationId)
        writer.string(2, url)
        writer.message(6, audioInfo.data)
        return writer.data
    }

    static func decodeTranslationResponse(_ data: Data) throws -> TranslationResponse {
        var reader = Reader(data)
        var result = TranslationResponse()

        while reader.hasRemaining {
            guard let tag = try reader.readTag() else { break }
            switch tag.fieldNumber {
            case 1: result.url = try reader.readString(wireType: tag.wireType)
            case 4: result.status = try reader.readInt32(wireType: tag.wireType)
            case 5: result.remainingTime = try reader.readInt32(wireType: tag.wireType)
            case 7: result.translationId = try reader.readString(wireType: tag.wireType)
            case 8: result.language = try reader.readString(wireType: tag.wireType)
            case 9: result.message = try reader.readString(wireType: tag.wireType)
            case 10: result.isLivelyVoice = try reader.readInt32(wireType: tag.wireType) != 0
            default: try reader.skip(wireType: tag.wireType)
            }
        }
        return result
    }

    private struct Tag {
        let fieldNumber: Int
        let wireType: Int
    }

    private struct Writer {
        var data = Data()

        mutating func int32(_ field: Int, _ value: Int) {
            guard value != 0 else { return }
            tag(field, wireVarint)
            varint(UInt64(UInt32(bitPattern: Int32(value))))
        }

        mutating func bool(_ field: Int, _ value: Bool) {
            guard value else { return }
            tag(field, wireVarint)
            varint(1)
        }

        mutating func double(_ field: Int, _ value: Double) {
            guard value != 0 else { return }
            tag(field, wireFixed64)
            var bits = value.bitPattern.littleEndian
            withUnsafeBytes(of: &bits) { data.append(contentsOf: $0) }
        }

        mutating func string(_ field: Int, _ value: String) {
            guard !value.isEmpty else { return }
            bytes(field, Data(value.utf8))
        }

        mutating func message(_ field: Int, _ value: Data) {
            bytes(field, value)
        }

        mutating func bytes(_ field: Int, _ value: Data) {
            tag(field, wireLengthDelimited)
            varint(UInt64(value.count))
            data.append(value)
        }

        mutating func tag(_ field: Int, _ wireType: Int) {
            precondition(field > 0)
            varint(UInt64((field << 3) | wireType))
        }

        mutating func varint(_ value: UInt64) {
            var value = value
            while true {
                if value & ~0x7f == 0 {
                    data.append(UInt8(value))
                    return
                }
                data.append(UInt8((value & 0x7f) | 0x80))
                value >>= 7
            }
        }
    }

    private struct Reader {
        let bytes: [UInt8]
        var position = 0

        init(_ data: Data) {
            self.bytes = Array(data)
        }

        var hasRemaining: Bool { position < bytes.count }

        mutating func readTag() throws -> Tag? {
            guard hasRemaining else { return nil }
            let raw = Int(try readVarint())
            guard raw != 0 else { return nil }
            return Tag(fieldNumber: raw >> 3, wireType: raw & 0x7)
        }

        mutating func readInt32(wireType: Int) throws -> Int {
            try requireWire(wireType, wireVarint)
            return Int(Int32(bitPattern: UInt32(truncatingIfNeeded: try readVarint())))
        }

        mutating func readString(wireType: Int) throws -> String {
            let data = try readBytes(wireType: wireType)
            guard let string = String(data: data, encoding: .utf8) else {
                throw ProtoError.malformed("Malformed UTF-8 in protobuf response")
            }
            return string
        }

        mutating func readBytes(wireType: Int) throws -> Data {
            try requireWire(wireType, wireLengthDelimited)
            let length = Int(try readVarint())
            guard length >= 0, position + length <= bytes.count else {
                throw ProtoError.malformed("Malformed protobuf length")
            }
            let result = Data(bytes[position..<(position + length)])
            position += length
            return result
        }

        mutating func skip(wireType: Int) throws {
            switch wireType {
            case wireVarint:
                _ = try readVarint()
            case wireFixed64:
                try advance(8)
            case wireLengthDelimited:
                try advance(Int(try readVarint()))
            case wireFixed32:
                try advance(4)
            default:
                throw ProtoError.malformed("Unsupported protobuf wire type: \(wireType)")
            }
        }

        mutating func advance(_ count: Int) throws {
            guard count >= 0, position + count <= bytes.count else {
                throw ProtoError.malformed("Malformed protobuf payload")
            }
            position += count
        }

        mutating func readVarint() throws -> UInt64 {
            var result: UInt64 = 0
            var shift: UInt64 = 0
            while shift < 64 {
                guard position < bytes.count else {
                    throw ProtoError.malformed("Truncated protobuf varint")
                }
                let byte = bytes[position]
                position += 1
                result |= UInt64(byte & 0x7f) << shift
                if byte & 0x80 == 0 { return result }
                shift += 7
            }
            throw ProtoError.malformed("Malformed protobuf varint")
        }

        func requireWire(_ actual: Int, _ expected: Int) throws {
            guard actual == expected else {
                throw ProtoError.malformed("Unexpected protobuf wire type: \(actual), expected \(expected)")
            }
        }
    }

    private static let wireVarint = 0
    private static let wireFixed64 = 1
    private static let wireLengthDelimited = 2
    private static let wireFixed32 = 5
}

File: app/src/main/java/com/github/libretube/vot/VotProto.kt
package com.github.libretube.vot

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Tiny protobuf wire codec for the handful of Yandex VOT messages we use.
 *
 * Keeping this codec local avoids adding a protobuf Gradle plugin/dependency to LibreTube.
 * Field numbers mirror the current vot.js schema (3.0.2, checked 2026-08-07).
 */
internal object VotProto {
    data class TranslationResponse(
        val url: String = "",
        val status: Int = 0,
        val remainingTime: Int = 0,
        val translationId: String = "",
        val language: String = "",
        val message: String = "",
        val isLivelyVoice: Boolean = false,
    )

    data class SessionResponse(
        val secretKey: String,
        val expires: Int,
    )

    fun encodeSessionRequest(uuid: String, module: String): ByteArray = Writer().apply {
        string(1, uuid)
        string(2, module)
    }.toByteArray()

    fun decodeSessionResponse(bytes: ByteArray): SessionResponse {
        val reader = Reader(bytes)
        var secretKey = ""
        var expires = 0
        while (reader.hasRemaining()) {
            val tag = reader.readTag() ?: break
            when (tag.fieldNumber) {
                1 -> secretKey = reader.readString(tag.wireType)
                2 -> expires = reader.readInt32(tag.wireType)
                else -> reader.skip(tag.wireType)
            }
        }
        require(secretKey.isNotBlank()) { "VOT session response did not include a secret key" }
        return SessionResponse(secretKey, expires)
    }

    fun encodeTranslationRequest(
        url: String,
        firstRequest: Boolean,
        duration: Double,
        language: String,
        responseLanguage: String,
        videoTitle: String,
        useLivelyVoice: Boolean = false,
    ): ByteArray = Writer().apply {
        string(3, url)
        bool(5, firstRequest)
        double(6, duration)
        int32(7, 1)
        string(8, language)
        string(14, responseLanguage)
        int32(15, 1)
        int32(16, 2)
        bool(18, useLivelyVoice)
        if (videoTitle.isNotBlank()) string(19, videoTitle)
    }.toByteArray()

    fun encodeAudioRequest(
        translationId: String,
        url: String,
        fileId: String,
    ): ByteArray {
        val audioInfo = Writer().apply {
            string(1, fileId)
        }.toByteArray()

        return Writer().apply {
            string(1, translationId)
            string(2, url)
            message(6, audioInfo)
        }.toByteArray()
    }

    fun decodeTranslationResponse(bytes: ByteArray): TranslationResponse {
        val reader = Reader(bytes)
        var url = ""
        var status = 0
        var remainingTime = 0
        var translationId = ""
        var language = ""
        var message = ""
        var isLivelyVoice = false

        while (reader.hasRemaining()) {
            val tag = reader.readTag() ?: break
            when (tag.fieldNumber) {
                1 -> url = reader.readString(tag.wireType)
                4 -> status = reader.readInt32(tag.wireType)
                5 -> remainingTime = reader.readInt32(tag.wireType)
                7 -> translationId = reader.readString(tag.wireType)
                8 -> language = reader.readString(tag.wireType)
                9 -> message = reader.readString(tag.wireType)
                10 -> isLivelyVoice = reader.readInt32(tag.wireType) != 0
                else -> reader.skip(tag.wireType)
            }
        }
        return TranslationResponse(url, status, remainingTime, translationId, language, message, isLivelyVoice)
    }

    private data class Tag(val fieldNumber: Int, val wireType: Int)

    private class Writer {
        private val out = ByteArrayOutputStream()

        fun int32(field: Int, value: Int) {
            if (value == 0) return
            tag(field, WIRE_VARINT)
            varint(value.toLong() and 0xffffffffL)
        }

        fun bool(field: Int, value: Boolean) {
            if (!value) return
            tag(field, WIRE_VARINT)
            varint(1)
        }

        fun double(field: Int, value: Double) {
            if (value == 0.0) return
            tag(field, WIRE_FIXED64)
            val bytes = ByteBuffer.allocate(8)
                .order(ByteOrder.LITTLE_ENDIAN)
                .putDouble(value)
                .array()
            out.write(bytes)
        }

        fun string(field: Int, value: String) {
            if (value.isEmpty()) return
            bytes(field, value.toByteArray(Charsets.UTF_8))
        }

        fun message(field: Int, value: ByteArray) = bytes(field, value)

        fun bytes(field: Int, value: ByteArray) {
            tag(field, WIRE_LENGTH_DELIMITED)
            varint(value.size.toLong())
            out.write(value)
        }

        private fun tag(field: Int, wireType: Int) {
            require(field > 0)
            varint(((field shl 3) or wireType).toLong())
        }

        private fun varint(value: Long) {
            var v = value
            while (true) {
                if ((v and -128L) == 0L) {
                    out.write(v.toInt())
                    return
                }
                out.write(((v and 0x7f) or 0x80).toInt())
                v = v ushr 7
            }
        }

        fun toByteArray(): ByteArray = out.toByteArray()
    }

    private class Reader(private val data: ByteArray) {
        private var position = 0

        fun hasRemaining(): Boolean = position < data.size

        fun readTag(): Tag? {
            if (!hasRemaining()) return null
            val raw = readVarint().toInt()
            if (raw == 0) return null
            return Tag(raw ushr 3, raw and 0x7)
        }

        fun readInt32(wireType: Int): Int {
            requireWire(wireType, WIRE_VARINT)
            return readVarint().toInt()
        }

        fun readString(wireType: Int): String = readBytes(wireType).toString(Charsets.UTF_8)

        private fun readBytes(wireType: Int): ByteArray {
            requireWire(wireType, WIRE_LENGTH_DELIMITED)
            val length = readVarint().toInt()
            require(length >= 0 && position + length <= data.size) { "Malformed protobuf length" }
            val result = data.copyOfRange(position, position + length)
            position += length
            return result
        }

        fun skip(wireType: Int) {
            when (wireType) {
                WIRE_VARINT -> readVarint()
                WIRE_FIXED64 -> advance(8)
                WIRE_LENGTH_DELIMITED -> advance(readVarint().toInt())
                WIRE_FIXED32 -> advance(4)
                else -> error("Unsupported protobuf wire type: $wireType")
            }
        }

        private fun advance(count: Int) {
            require(count >= 0 && position + count <= data.size) { "Malformed protobuf payload" }
            position += count
        }

        private fun readVarint(): Long {
            var result = 0L
            var shift = 0
            while (shift < 64) {
                require(position < data.size) { "Truncated protobuf varint" }
                val b = data[position++].toInt() and 0xff
                result = result or ((b and 0x7f).toLong() shl shift)
                if ((b and 0x80) == 0) return result
                shift += 7
            }
            error("Malformed protobuf varint")
        }

        private fun requireWire(actual: Int, expected: Int) {
            require(actual == expected) { "Unexpected protobuf wire type: $actual, expected $expected" }
        }
    }

    private const val WIRE_VARINT = 0
    private const val WIRE_FIXED64 = 1
    private const val WIRE_LENGTH_DELIMITED = 2
    private const val WIRE_FIXED32 = 5
}
